pkg/chains/formats/intotoite6/util/utils.go

```diff
@@ -93,3 +94,64 @@ func SpdxGit(url, revision string) string {
 	}
 	return prefix + url + fmt.Sprintf("@%s", revision)
 }
+
+type StepAttestation struct {
+	EntryPoint  string            `json:"entryPoint"`
+	Arguments   interface{}       `json:"arguments,omitempty"`
+	Environment interface{}       `json:"environment,omitempty"`
+	Annotations map[string]string `json:"annotations"`
+}
+
+func AttestStep(step *v1beta1.Step, stepState *v1beta1.StepState) StepAttestation {
+	attestation := StepAttestation{}
+
+	entrypoint := strings.Join(step.Command, " ")
+	if step.Script != "" {
+		entrypoint = step.Script
+	}
+	attestation.EntryPoint = entrypoint
+	attestation.Arguments = step.Args
+
+	env := map[string]interface{}{}
+	env["image"] = stepState.ImageID
+	env["container"] = stepState.Name
+	attestation.Environment = env
+
+	return attestation
+}
+
+func AttestInvocation(params []v1beta1.Param, paramSpecs []v1beta1.ParamSpec) slsa.ProvenanceInvocation {
+	i := slsa.ProvenanceInvocation{}
+	iParams := make(map[string]string)
+
+	// get implicit parameters from defaults
+	for _, p := range paramSpecs {
+		if p.Default != nil {
+			// TODO: Consider using p.Default.MarshalJSON()
+			var v string
+			switch p.Default.Type {
+			case v1beta1.ParamTypeString:
+				v = p.Default.StringVal
+			case v1beta1.ParamTypeArray:
+				v = fmt.Sprintf("%v", p.Default.ArrayVal)
+			}
+			iParams[p.Name] = v
```

Comment:
Yeah, I'd go with `MarshalJSON`, with `%v` values would be separated by space so if a value contains a space it would be difficult to tell the values apart. Would this work?

```suggestion
			if bytes, err := p.Default.MarshalJSON(); err == nil {
				iParams[p.Name] = string(bytes)
			} else {
				panic(err) // or change the signature to also return error
			}
```

Reply:
It would. I'm asking upstream if I should just switch to MarshalJSON. I can't imagine they'd object.

Maybe panic is not ideal here. I definitely don't want the controller to crash due to some unforeseen circumstance. Log an error and ignore the param is probably the best approach here.